crates/iceberg/src/spec/view_metadata_builder.rs

```diff
@@ -0,0 +1,1579 @@
+            .metadata
+            .properties
+            .get(VIEW_PROPERTY_VERSION_HISTORY_SIZE)
+            .and_then(|v| v.parse::<usize>().ok())
+            .unwrap_or(VIEW_PROPERTY_VERSION_HISTORY_SIZE_DEFAULT)
+            .max(1);
+
+        // expire old versions, but keep at least the versions added in this builder
+        let num_added_versions = self
+            .changes
+            .iter()
+            .filter(|update| matches!(update, ViewUpdate::AddViewVersion { .. }))
+            .count();
+        let num_versions_to_keep = num_added_versions.max(num_versions_to_keep);
+
+        if self.metadata.versions.len() > num_versions_to_keep {
+            // version ids are assigned sequentially. keep the latest versions by ID.
+            let mut versions_to_keep = self
+                .metadata
+                .versions
+                .keys()
+                .copied()
+                .sorted()
+                .rev()
+                .take(num_versions_to_keep)
+                .collect::<HashSet<_>>();
+
+            // always retain current version
+            if !versions_to_keep.contains(&self.metadata.current_version_id) {
+                // Remove the lowest ID
+                if num_versions_to_keep > num_added_versions {
+                    let lowest_id = versions_to_keep.iter().min().copied();
+                    lowest_id.map(|id| versions_to_keep.remove(&id));
+                }
+                // Add the current version ID
+                versions_to_keep.insert(self.metadata.current_version_id);
+            }
+
+            let mut expired_versions = Vec::new();
+            // remove all versions which are not in versions_to_keep from the metadata
+            // and add them to the expired_versions list
+            self.metadata.versions.retain(|id, version| {
+                if versions_to_keep.contains(id) {
+                    true
+                } else {
+                    expired_versions.push(version.clone());
+                    false
+                }
+            });
+
+            expired_versions
+        } else {
+            Vec::new()
+        }
+    }
+}
+
+/// Expire version log entries that are no longer relevant.
+/// Returns the history entries to retain.
+fn update_version_log(
+    version_log: Vec<ViewVersionLog>,
+    ids_to_keep: HashSet<i32>,
+) -> Vec<ViewVersionLog> {
+    let mut retained_history = Vec::new();
+    for log_entry in version_log {
+        if ids_to_keep.contains(&log_entry.version_id()) {
+            retained_history.push(log_entry);
+        } else {
+            retained_history.clear();
+        }
+    }
+    retained_history
+}
+
+fn allow_replace_drop_dialects(properties: &HashMap<String, String>) -> bool {
+    properties
+        .get(VIEW_PROPERTY_REPLACE_DROP_DIALECT_ALLOWED)
+        .map_or(
+            VIEW_PROPERTY_REPLACE_DROP_DIALECT_ALLOWED_DEFAULT,
+            |value| value.to_lowercase() == "true",
```

Comment:
Would it make sense to move this into some utility method?

Reply:
I think we should revisit properties for views and tables together.
I think I'll give it a shot next week.